dao/src/main/java/com/epam/esm/dao/repository/impl/UserRepositoryImpl.java

```diff
@@ -0,0 +1,61 @@
+package com.epam.esm.dao.repository.impl;
+
+import com.epam.esm.dao.entity.User;
+import com.epam.esm.dao.repository.UserRepository;
+import com.epam.esm.dao.specification.Specification;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.persistence.EntityManager;
+import javax.persistence.NoResultException;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+public class UserRepositoryImpl implements UserRepository {
+  private final EntityManager entityManager;
+
+  @Autowired
+  public UserRepositoryImpl(EntityManager entityManager) {
+    this.entityManager = entityManager;
+  }
+
+  @Override
+  public List<User> getEntityListBySpecification(Specification<User> specification) {
+    CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+    CriteriaQuery<User> criteriaQuery = specification.getCriteriaQuery(builder);
+    return entityManager.createQuery(criteriaQuery).getResultList();
+  }
+
+  @Override
+  public List<User> getEntityListWithPaginationBySpecification(
```

Comment:
Actually I think this is redundant to add 'BySpecfication' or 'WithPagination'. This method could be just getListBy...
And some explanations may be added in scope of Java Doc

Reply:
Yes, you're absolutely right. (fixed)